cpp/src/link_prediction/similarity_impl.cuh

```diff
@@ -162,5 +163,390 @@ rmm::device_uvector<weight_t> similarity(
+
+    if (vertices) {
+      tmp_vertices.resize(vertices->size(), handle.get_stream());
+      thrust::copy(
+        handle.get_thrust_policy(), vertices->begin(), vertices->end(), tmp_vertices.begin());
+    } else {
+      tmp_vertices.resize(graph_view.local_vertex_partition_range_size(), handle.get_stream());
+      thrust::sequence(handle.get_thrust_policy(),
+                       tmp_vertices.begin(),
+                       tmp_vertices.end(),
+                       graph_view.local_vertex_partition_range_first());
+    }
+
+    //  We can reduce memory footprint by doing work in batches and
+    //  computing/updating topk with each batch
+
+    //   FIXME: Experiment with this and adjust as necessary
+    // size_t const
+    // MAX_PAIRS_PER_BATCH{static_cast<size_t>(handle.get_device_properties().multiProcessorCount) *
+    // (1 << 15)};
+    size_t const MAX_PAIRS_PER_BATCH{100};
+
+    rmm::device_uvector<edge_t> degrees = graph_view.compute_out_degrees(handle);
+    rmm::device_uvector<size_t> two_hop_degrees(degrees.size() + 1, handle.get_stream());
+
+    // Let's compute the maximum size of the 2-hop neighborhood of each vertex
+    // FIXME: If vertices is specified, this could be done on a subset of the vertices
+    //
+    edge_dst_property_t<GraphViewType, edge_t> edge_dst_degrees(handle, graph_view);
+    update_edge_dst_property(handle, graph_view, degrees.begin(), edge_dst_degrees);
+
+    per_v_transform_reduce_incoming_e(
+      handle,
+      graph_view,
+      edge_src_dummy_property_t{}.view(),
+      edge_dst_degrees.view(),
+      edge_dummy_property_t{}.view(),
+      [] __device__(vertex_t, vertex_t, auto, auto dst_degree, auto) {
+        return static_cast<size_t>(dst_degree);
+      },
+      size_t{0},
+      reduce_op::plus<size_t>{},
+      two_hop_degrees.begin());
+
+    if (vertices) {
+      rmm::device_uvector<size_t> gathered_degrees(tmp_vertices.size() + 1, handle.get_stream());
+
+      thrust::gather(
+        handle.get_thrust_policy(),
+        thrust::make_transform_iterator(
+          tmp_vertices.begin(),
+          cugraph::detail::shift_left_t<vertex_t>{graph_view.local_vertex_partition_range_first()}),
+        thrust::make_transform_iterator(
+          tmp_vertices.end(),
+          cugraph::detail::shift_left_t<vertex_t>{graph_view.local_vertex_partition_range_first()}),
+        two_hop_degrees.begin(),
+        gathered_degrees.begin());
+
+      two_hop_degrees = std::move(gathered_degrees);
+    }
+
+    thrust::sort_by_key(handle.get_thrust_policy(),
+                        two_hop_degrees.begin(),
+                        two_hop_degrees.end() - 1,
+                        tmp_vertices.begin(),
+                        thrust::greater<size_t>{});
+
+    thrust::exclusive_scan(handle.get_thrust_policy(),
+                           two_hop_degrees.begin(),
+                           two_hop_degrees.end(),
+                           two_hop_degrees.begin());
```

Comment:
auto two_hop_degree_offsets = std::move(two_hop_degrees)? `two_hop_degrees` is a misnomer from this point.

Reply:
OK.  I added this.  Don't use it very much from here, but it's essentially free to move and perhaps a bit clearer.